### PR TITLE
Deprecate using 0 value for `min_children` in `has_child` query #41548

### PR DESCRIPTION
**Pull request by codebird**
Added deprecation warning when min_children is 0

Closes https://github.com/elastic/elasticsearch/issues/41548

### CODE REVIEW

**Comment by codebird** [comment]
related to #41548

**Comment by codebird** [comment]
@cbuescher I think I covered everything. Please review.

File: modules/parent-join/src/test/java/org/elasticsearch/join/query/HasChildQueryBuilderTests.java [hunk 1]
```diff
@@ -163,6 +163,13 @@ protected HasChildQueryBuilder doCreateTestQueryBuilder() {
         return hqb;
     }
 
+    public void testZeroMinChildren() {
```
Comment (dwlizlo) [inline modules/parent-join/src/test/java/org/elasticsearch/join/query/HasChildQueryBuilderTests.java]:
```suggestion
    public void testDeprecationOfZeroMinChildren() {
```

**Comment by elasticcla** [comment]
Hi @codebird, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in your Git [commit](https://github.com/elastic/elasticsearch/pull/41555.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java [hunk 2]
```diff
@@ -136,6 +140,9 @@ public HasChildQueryBuilder minMaxChildren(int minChildren, int maxChildren) {
         if (minChildren < 0) {
             throw new IllegalArgumentException("[" + NAME + "] requires non-negative 'min_children' field");
         }
+        if (minChildren == 0) {
```
Comment (cbuescher) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: Since maxChildren == 0 is also disallowed in 8.0, that should get a warning too.

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java [hunk 3]
```diff
@@ -66,21 +68,23 @@
...
-    public static final int DEFAULT_MIN_CHILDREN = 0;
+    public static final int DEFAULT_MIN_CHILDREN = 1;
 
     /**
      * The default value for ignore_unmapped.
      */
     public static final boolean DEFAULT_IGNORE_UNMAPPED = false;
-
+    static final String MIN_CHILDREN_0_DEPRECATION_MESSAGE = "[min_children] 0 will be rejected " +
+        "in 8.0, use 1 and your query results won't change.";
```
Comment (cbuescher) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: nit: "... starting with 8.0". 
And maybe rephrase the last part a little more formally, e.g. "Using 1 instead of 0 will return the same result set".

**Review by cbuescher — CHANGES_REQUESTED** [review]
@codebird thanks, looks pretty complete already. I left a few very small requests for changes, but those shouldn't be too involved.

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java (hunk 2 above)
Comment (codebird) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: @cbuescher do you really think we need this warning? if maxChildren is 0 then either minChildren will be 0 so we will get a warning for it, or maxChildren < minChildren will throw an error. That's what I thought, what do you think?

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java (hunk 3 above)
Comment (codebird) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: this is done, will push when we figure out the above.

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java (hunk 2 above)
Comment (codebird) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: I updated DEFAULT_MIN_CHILDREN to be 1 at the place of 0. So the case where min_children is not provided but max_children is set to 0, this throws an error. Do you think I should keep DEFAULT_MIN_CHILDREN 0, this will cause CI to fail because they'll all be throwing warnings?

Comment (codebird) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: So basically we have 2 solutions here:
1- Put DEFAULT_MIN_CHILDREN back to 0, and fix all CI tests not to cause the warnings, and add the warning for maxChildren == 0

2- Keep DEFAULT_MIN_CHILDREN 1, at the risk of breaking queries for people who might be providing a query like this 1:
```
{
  "query": {
    "has_child": {
      "type":"answer",
      "max_children": 0, 
      "query": {
        "match_all": {}
      }
    }
  }
}'
```

Comment (codebird) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: one last detail to consider, is that if we go with solution 1, the deprecation warning will be displayed when the query doesn't have min_children, which will cause the user to think that min_children is mandatory, which is not the case. (that's why I updated DEFAULT_MIN_CHILDREN). sorry for spamming :) I am brainstorming here

Comment (cbuescher) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: Thanks for the explanation, I think you are right in updating the  DEFAULT_MIN_CHILDREN and this will give sufficient warning if we only check the min value, otherwise the later maxChildren < minChildren condition will trigger an error. The case where "minChildren" is unset and "maxChildren" set to 0 is highly unlikely and can be neglected I think.

**Comment by codebird** [comment]
@cbuescher Error message updated as you asked.

**Review by cbuescher — APPROVED** [review]
Thanks @codebird, lets start the CI dance ;-)

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by codebird** [comment]
That's just not to change the habits before the weekend

**Comment by cbuescher** [comment]
@elasticmachine run  elasticsearch-ci/1

**Comment by cbuescher** [comment]
Looks like we're in luck this time. Will merge to 7.x then. Thanks again @codebird

**Comment by codebird** [comment]
Thanks for the support @cbuescher
